.changelog/unreleased/breaking-changes/1203-bytes-for-abci-proto.md

```diff
@@ -0,0 +1,2 @@
+- `[tendermint-proto]` Use `Bytes` for byte array fields of ABCI protobuf types.
+  ([#1203](https://github.com/informalsystems/tendermint-rs/pull/1203))
```

Comment:
@thanethomson I've added a notice about this change, but as stated already, I'm not sure we need to go ahead with it for 0.26. This is certainly unrelated to domain types, applied piecemeal to a single proto module not as a consistent change throughout tendermint-proto, and has the overall smell of a premature optimization.

Reply:
This is related to the domain types, because the ABCI domain types use `Bytes` internally, so unwinding this part of the ABCI domain types will break existing users of the ABCI domain types. This is not meaningfully a breaking change for anyone else, because there weren't users of the ABCI proto types other than the stub ABCI demo application.  It's also not a breaking change relative to `master`, because this code has been merged on that branch (albeit unreleased) for about a year.

If this change is unwound out, you'll have to edit all of the ABCI domain types, and then all the users will have to edit all their uses of those domain types. I don't think it would be good to put this on the critical path for doing a semver release of the `tendermint` crates that contains ABCI domain types.

If the decision, later, is that the domain types should **not** use the `Bytes` type, it would be much better to do that as a later change that can be modeled with semver.  (For the record, I don't think this is a good decision, but in any case I think it should be separate from getting this code backported as-is).

Reply:
Thank you @hdevalence for the detailed exposition.

As a late-to-the-scene comment, if I were into micro-optimizing domain type conversions to protobuf, I'd start with the `Protobuf` trait forcing a clone of the entire structure, because the domain-to-raw conversion is implemented via `From<Self>`, but the `encode*` methods take `&self`.